quinn-proto/src/connection.rs

```diff
@@ -2847,6 +2888,57 @@ where
         Ok(n)
     }
 
+    /// Transmit an unreliable, unordered datagram
+    ///
+    /// `flow_id` is an arbitrary application-controlled value. If set to 0, it will consume no
+    /// additional space in the packet.
+    pub fn send_datagram(&mut self, data: Bytes) -> Result<(), SendDatagramError> {
+        if self.config.datagram_window.is_none() {
+            return Err(SendDatagramError::Disabled);
+        }
+        if data.len()
+            > self
+                .datagram_mtu()
+                .ok_or(SendDatagramError::UnsupportedByPeer)?
+        {
+            return Err(SendDatagramError::TooLarge);
+        }
+        if self.congestion_blocked() {
+            return Err(SendDatagramError::Blocked);
+        }
+        self.out_datagrams.push_back(Datagram { data });
+        Ok(())
+    }
+
+    /// Receive an unreliable, unordered datagram
+    pub fn recv_datagram(&mut self) -> Option<Bytes> {
+        let x = self.in_datagrams.pop_front()?.data;
+        self.datagram_bytes -= x.len();
+        Some(x)
+    }
+
+    /// Compute the maximum size of datagrams that may be sent, if supported on this connection
+    ///
+    /// This may change over the lifetime of a connection according to variation in the MTU
+    /// estimate, but will never be below the smaller of `MIN_DATAGRAM_MTU` and the limit specified
+    /// by the peer.
+    pub fn datagram_mtu(&self) -> Option<usize> {
+        self.config.datagram_window?;
+        let limit = self.params.max_datagram_frame_size?.into_inner();
+        Some(limit.min(self.max_payload_size() as u64 - Datagram::SIZE_BOUND as u64) as usize)
+    }
```

Comment:
I'm particularly interested in feedback on this function, and the related constant. It strikes me as potentially confusing, but the only other idea I have so far is exposing the components--peer limit and path MTU--separately, which is error-prone in that users might neglect to check both, and doesn't have an obvious functional benefit.

Reply:
Can you link to relevant part(s) of the draft?

Reply:
I don't understand what the call to `into_inner()` is doing. As I understand it, `self.params.max_datagram_frame_size?` is `usize`, so what's the `into_inner()` doing here? Also let's inline `max_payload_size()`, since it's a single-use function (and the name doesn't suggest it is only relevant for datagrams). Also, "will never be below ... the limit specified by the peer" doesn't seem very meaningful or helpful since the limit specified by the peer, as I understand it, might be 1? Also, the name here ending in `_mtu` again suggests an MTU unit, whereas it's actually datagram payload size?